sql/2019/12_Mobile_Web/12_13.sql

```diff
@@ -0,0 +1,24 @@
+#standardSQL
+
+# input types occurence prefined set %
+
+CREATE TEMPORARY FUNCTION getInputTypes(payload STRING)
+RETURNS ARRAY<STRING> LANGUAGE js AS '''
+  try {
+    var $ = JSON.parse(payload);
+    var almanac = JSON.parse($._almanac);
+    return almanac['input-elements'] && almanac['input-elements'].map(function(node) {
+        return node.type.toLowerCase();
+    });
+  } catch (e) {
+    return [];
+  }
+''';
+
+SELECT
+    COUNTIF(REGEXP_CONTAINS(input_types, '/color|date|datetime-local|email|month|number|range|reset|search|tel|time|url|week|datalist/i')) AS occurence,
+    ROUND(COUNT(REGEXP_CONTAINS(input_types, '/color|date|datetime-local|email|month|number|range|reset|search|tel|time|url|week|datalist/i')) * 100 / SUM(COUNT(0)) OVER (), 2) AS occurence_perc
```

Comment:
This is the % of input types that are in the list. I think what the metric is asking for is the % of _websites_ having any of the types in the list. Here's an example of doing that logic in the UDF:

```sql
#standardSQL

# input types occurence prefined set %

CREATE TEMPORARY FUNCTION hasNewInputType(payload STRING)
RETURNS BOOLEAN LANGUAGE js AS '''
  try {
    var $ = JSON.parse(payload);
    var almanac = JSON.parse($._almanac);
    var newTypes = new Set(['color', 'date', 'datetime-local', 'email', 'month', 'number', 'range', 'reset', 'search', 'tel', 'time', 'url', 'week', 'datalist']);
    return !!almanac['input-elements'].find(node => newTypes.has(node.type.toLowerCase()));
  } catch (e) {
    return false;
  }
''';

SELECT
    COUNTIF(hasNewInputType(payload)) AS occurence,
    ROUND(COUNTIF(hasNewInputType(payload)) * 100 / SUM(COUNT(0)) OVER (), 2) AS occurence_perc
FROM
    `httparchive.pages.2019_07_01_mobile`
```